### ISSUE TITLE
Parse full description from ASHP shortage detail page

### ISSUE THREAD

**Issue by jrlegrand** [coderxio/sagerx#325]
# Problem Statement
In addition to getting the NDC from this, also grab the full bullet point of text and associate it with the NDC.

Some parts like pack size would be useful to pull out by exploding by commas and comparing to other sources of pack size.

<img width="773" alt="image" src="https://github.com/user-attachments/assets/3370e9c0-fb04-4b3f-bd56-6f74255f65ec">

# Criteria for Success
For both affected and available NDCs, add a column in the ndc model for "description" or something.

# Additional Information
Needs to be in the bs4 code, plus an additional column in the dbt model.

**Comment by eddie-cosma** [comment]
I've been playing around with this. I can split each line by the comma character but the problem is that the resulting list length is variable from one bullet point to another. It seems like the general structure is something like this:

```
0.9% Sodium Chloride Irrigation, BBraun, 1000 mL bottle, 16 count, NDC 00264-2201-00
| ---------- product --------- |  mfg  | - package description - | ------ NDC ----- |
```

Splitting out the product and manufacturer at the beginning, and the NDC at the end, means that the package description section makes up the remaining data, regardless of how many commas are in there.

In practice, this looks something like this:

<img width="1143" alt="Image" src="https://github.com/user-attachments/assets/d2144b23-9902-4f23-bd1e-f38a5e195659" />

<img width="1234" alt="Image" src="https://github.com/user-attachments/assets/48cd8535-572f-4d7c-811f-19bf4d9c3945" />

Before I move on to updating the dbt model, are these descriptions actually useful for us or do we need to split this out further?

**Comment by jrlegrand** [comment]
Ah - i see what you mean.  i think this is perfect.

I can pull that description into another report i'm working on to help me validate it.

Are you storing the full bullet point via scraping and then chopping it up with dbt SQL?

**Comment by eddie-cosma** [comment]
Currently doing the splitting in Python before storing anything:

```python
        for ndc_detail in soup.find(id='1_lblProducts').find_all('li'):
            product, manufacturer, *description, ndc = ndc_detail.get_text().split(',')
            description = ','.join(description)
            ndc_data = {
                'product': product,
                'manufacturer': manufacturer,
                'description': description,
                'ndc': ndc.replace('NDC ', '').replace('-', ''),
                'detail_url': shortage['detail_url'],
                'ndc_type': 'affected'
            }
            affected_ndcs.append(ndc_data)
```

Then using Pandas.json_normalize() to turn it into an appropriate DataFrame.

Could be done in SQL also but Python was more convenient for a proof of concept. Any preference between the two?

**Comment by jrlegrand** [comment]
My preference is to preserve the original data with as much fidelity as possible as part of the extract / load steps and try to keep transformation in SQL (dbt) as much as possible. This allows us to change how we transform the data without having to reload the data.

In short, I'd prefer to just pull down and store the entire bulleted sentence using bs4 and then split it using SQL via a dbt staging model.

I'm happy to help orient you to dbt by pushing a commit to show the SQL I would write and where I would put it. Or if you want to write the SQL, I can just demonstrate where I would put it as part of a dbt model.

**Comment by eddie-cosma** [comment]
Makes sense. The SQL should be pretty easy. I'll work on it and let you know if I encounter any problems.

**Comment by eddie-cosma** [comment]
I think I have this mostly cleaned up. Here's a sampling of data from the updated shortage NDC table:

| id | product | manufacturer | description | ndc\_11 | ndc\_type |
| :--- | :--- | :--- | :--- | :--- | :--- |
| 1091 | Acamprosate calcium oral tablet |  Glenmark |  333 mg, bottle, 180 count | 68462043518 | affected |
| 977 | Theophylline extended release tablet |  Glenmark |  400 mg, bottle, 100 count | 68462038001 | available |
| 977 | Theophylline extended release tablet |  Glenmark |  600 mg, bottle, 100 count | 68462035601 | available |
| 1091 | Acamprosate calcium oral tablet |  Zydus |  333 mg, bottle, 180 count | 68382056928 | affected |
| 953 | Ribavirin inhalation powder for solution |  Zydus |  6 gram, 100 mL vial, 4 count | 68382039504 | available |
| 675 | Desmopressin Acetate nasal spray |  Zydus |  10 mcg/0.1 mL, 5 mL bottle | 68382038401 | affected |
| 956 | Budesonide inhalation suspension |  Lupin |  0.5 mg / 2 mL, 5 single-dose ampules per pouch, 6 pouches, 30 count | 68180098430 | available |

I did find two edge cases:

[This shortage](https://www.ashp.org/drug-shortages/current-shortages/drug-shortage-detail.aspx?id=859) does not contain an NDC in the sentence:

<img width="804" alt="Image" src="https://github.com/user-attachments/assets/2279e321-a968-44fb-b270-2912fe914a48" />

It renders like so:
| id | product | manufacturer | description | ndc\_11 | ndc\_type |
| :--- | :--- | :--- | :--- | :--- | :--- |
| 859 | Magnesium citrate oral solution |  Vi-Jon |  1.745 gm/30 mL, 10 ounce bottle | _null_ | affected |

Additionally, [this shortage](https://www.ashp.org/drug-shortages/current-shortages/drug-shortage-detail.aspx?id=847) references an imported product:

<img width="808" alt="Image" src="https://github.com/user-attachments/assets/6629bc98-0ef1-4ace-980f-f3906d884e66" />

It renders like so:
| id | product | manufacturer | description | ndc\_11 | ndc\_type |
| :--- | :--- | :--- | :--- | :--- | :--- |
| 847 | Imported non-FDA-approved Zanosar 1 gram vials |  Marketing Authorization Number: PL 40308-0001 |  | _null_ | available |

I'm a little torn between keeping these (since they do link back to the shortage ID and therefore add context) versus simply removing lines where `ndc_11` is null (since this is an NDC table after all). Any thoughts?

**Comment by jrlegrand** [comment]
I vote keep them even if they don't have an NDC. Makes most sense for a staging table. We can remove the null NDCs downstream if we want, but I think maintaining the same fidelity of information as the source in a staging table is good.

Looking good!

**Comment by jrlegrand** [comment]
Also, think about a data mart that would be useful now that all this data is going to be automatically flowing into staging tables like this. We could aggregate NDCs by RxNorm product-level RXCUI (would that be useful?) or maybe by a certain class in RxClass (like ATC) to see which class of drugs has the most current shortages. Things like that. You might have much better ideas than these. Just a jumping off point.

**Comment by jrlegrand** [comment]
One thing I seem to recall is that these shortage alerts don't represent ALL NDCs for the same product-level RXCUI. In other words, the "shortage" and "available" NDCs don't add up to all possible NDCs for a given type of product. It would be interesting to do some analysis on this and see if we could produce an "extended" shortage prediction based on other NDCs out there.

**Comment by eddie-cosma** [comment]
Good call. I will do some brainstorming to see how we can incorporate all this into a data mart. In the meantime, I created a PR to at least get the new data live.
